src/site/content/en/blog/webassembly-memory-debugging/index.md

```diff
@@ -0,0 +1,607 @@
+
+  module.free_result();
+  module.doLeakCheck();
+
+  return new ImageData(
+    new Uint8ClampedArray(result.view),
+    result.width,
+    result.height
+  );
+}
+```
+
+This gives us a report like the following in the console:
+
+![Screenshot of a message "Direct leak of 12 bytes" coming from an anonymous wasm-function[…]](leak1-nodbg.png)
+
+Uh-oh, there are some small leaks, but the stacktrace is not very helpful as all the function names
+are mangled. Let's recompile with a basic debugging info to preserve them:
+
+```shell/12/
+emcc \
+  --bind \
+  ${OPTIMIZE} \
+  --closure 1 \
+  -s ALLOW_MEMORY_GROWTH=1 \
+  -s MODULARIZE=1 \
+  -s 'EXPORT_NAME="imagequant"' \
+  -I node_modules/libimagequant \
+  -o ./imagequant.js \
+  --std=c++11 \
+  imagequant.cpp \
+  -fsanitize=address \
+  -g2 \
+  node_modules/libimagequant/libimagequant.a
+```
+
+This looks much better:
+
+![Screenshot of a message "Direct leak of 12 bytes" coming from a `GenericBindingType<RawImage>::toWireType` function](leak1-dbg.png)
+
+Some parts of the stacktrace still look obscure as they point to Emscripten internals, but we can
+tell that the leak is coming from a `RawImage` conversion to "wire type" (to a JavaScript value) by
+Embind. Indeed, when we look at the code, we can see that we return `RawImage` C++ instances to
+JavaScript, but we never free them on either side.
+
+As a reminder, currently there is no garbage collection integration between JavaScript and
+WebAssembly, although [one is being developed](https://github.com/WebAssembly/gc). Instead, you have
+to manually free any memory and call destructors from the JavaScript side once you're done with the
+object. For Embind specifically, the [official
+docs](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#memory-management)
+suggest to call a `.delete()` method on exposed C++ classes:
+
+> JavaScript code must explicitly delete any C++ object handles it has received, or the Emscripten
+> heap will grow indefinitely.
+>
+> ```js
+> var x = new Module.MyClass;
+> x.method();
+> x.delete();
+> ```
+
+Indeed, when we do that in JavaScript for our class:
+
+```ts/7/
+  // …
+
+  const result = opts.zx
+    ? module.zx_quantize(data.data, data.width, data.height, opts.dither)
+    : module.quantize(data.data, data.width, data.height, opts.maxNumColors, opts.dither);
+
+  module.free_result();
+  result.delete();
+  module.doLeakCheck();
+
+  return new ImageData(
+    new Uint8ClampedArray(result.view),
+    result.width,
+    result.height
+  );
+}
+```
+
+The leak goes away as expected.
+
+### Discovering more issues with sanitizers
+
+Building other Squoosh codecs with sanitizers reveals both similar as well as some new issues. For
+example, I've got this error in MozJPEG bindings:
+
+![Screenshot of a message "memory access out of bounds" coming from a `jpeg_start_compress` function](out-of-bounds.png)
```

Comment:
Please use this formatting: https://web.dev/handbook/markup-media/#partial-screenshots